id_minter/src/main/scala/uk/ac/wellcome/platform/idminter/steps/IdentifierGenerator.scala

```diff
@@ -1,73 +1,76 @@
-            }
+  def retrieveOrGenerateCanonicalId(identifiers: List[SourceIdentifier],
+                                    ontologyType: String): Try[String] = {
+    Try {
+      val idsWithKnownSchemes = identifiers.filter(identifier =>
+        knownIdentifierSchemeList.contains(identifier.identifierScheme))
+      if (idsWithKnownSchemes.isEmpty) {
+        throw UnableToMintIdentifierException(
+          "identifiers list did not contain a known identifierScheme")
+      } else {
+        identifiersDao.lookupID(idsWithKnownSchemes, ontologyType).flatMap {
+          case Some(id) =>
+            metricsSender.incrementCount("found-old-id")
+            Try(id.CanonicalID)
           case None =>
-            error(s"Item $work did not contain a MiroID")
-            Future.failed(new Exception(s"Item $work did not contain a MiroID"))
-      }
-    )
-  }
-
-  private def retrieveOrGenerateCanonicalId(
-    identifier: SourceIdentifier,
-    ontologyType: String): Try[String] = {
-    identifiersDao.lookupID(List(identifier), ontologyType).flatMap {
-      case Some(id) =>
-        metricsSender.incrementCount("found-old-id")
-        Try(id.CanonicalID)
-      case None =>
-        val result = generateAndSaveCanonicalId(identifier.value)
-
-        metricsSender.incrementCount("generated-new-id")
+            val result =
+              generateAndSaveCanonicalId(idsWithKnownSchemes, ontologyType)
+            if (result.isSuccess)
+              metricsSender.incrementCount("generated-new-id")
 
-        result
-    }
-  }
-
-  private def findMiroID(work: Work): Option[SourceIdentifier] = {
-    val maybeSourceIdentifier =
-      work.identifiers.find(identifier =>
-        identifier.identifierScheme == IdentifierSchemes.miroImageNumber)
-    info(s"SourceIdentifier: $maybeSourceIdentifier")
-    maybeSourceIdentifier
+            result
+        }
+      }
+    }.flatten
   }
 
-  private def generateAndSaveCanonicalId(miroId: String): Try[String] = {
+  private def generateAndSaveCanonicalId(identifiers: List[SourceIdentifier],
+                                         ontologyType: String): Try[String] = {
     val canonicalId = Identifiable.generate
     identifiersDao
-      .saveIdentifier(Identifier(MiroID = miroId, CanonicalID = canonicalId))
+      .saveIdentifier(
+        Identifier(
+          MiroID =
+            findIdentifierWith(identifiers, IdentifierSchemes.miroImageNumber),
+          CalmAltRefNo =
+            findIdentifierWith(identifiers, IdentifierSchemes.calmAltRefNo),
+          CanonicalID = canonicalId,
+          ontologyType = ontologyType
+        ))
       .map { _ =>
         canonicalId
       }
   }
+
+  private def findIdentifierWith(identifiers: List[SourceIdentifier],
+                                 identifierScheme: String): String = {
+    identifiers
+      .find(identifier => identifier.identifierScheme == identifierScheme)
+      .fold[String](null)(identifier => identifier.value)
```

Comment:
Is `null` what we want here? Can we use `None`?

Reply:
Yeah.. this is a slightly unfortunate thing about how the table is defined for scalalikejdbc..